rust/src/derivation.rs

```diff
@@ -0,0 +1,539 @@
+        let ua_str = unsafe { CStr::from_ptr(ua).to_str()? };
+
+        let (network, ua) = match ZcashAddress::try_from_encoded(ua_str) {
+            Ok(addr) => addr
+                .convert::<(_, UnifiedAddressParser)>()
+                .map_err(|e| anyhow!("Not a Unified Address: {}", e)),
+            Err(e) => return Err(anyhow!("Invalid Zcash address: {}", e)),
+        }?;
+
+        if let Some(addr) = ua.0.sapling() {
+            Ok(
+                CString::new(ZcashAddress::from_sapling(network, addr.to_bytes()).encode())?
+                    .into_raw(),
+            )
+        } else {
+            Err(anyhow!(
+                "Unified Address doesn't contain a Sapling receiver"
+            ))
+        }
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the "wallet level", i.e.
+/// directly from the seed with no ZIP 32 path applied.
+///
+/// The resulting key will be the same across all networks (Zcash mainnet, Zcash testnet, OtherCoin
+/// mainnet, and so on). You can think of it as a context-specific seed fingerprint that can be used
+/// as (static) key material.
+///
+/// `context_string` is a globally-unique non-empty sequence of at most 252 bytes that identifies
+/// the desired context.
+///
+/// # Safety
+///
+/// - `context_string` must be non-null and valid for reads for `context_string_len` bytes, and it
+///   must have an alignment of `1`.
+/// - The memory referenced by `context_string` must not be mutated for the duration of the function
+///   call.
+/// - The total size `context_string_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - `seed` must be non-null and valid for reads for `seed_len` bytes, and it must have an
+///   alignment of `1`.
+/// - The memory referenced by `seed` must not be mutated for the duration of the function call.
+/// - The total size `seed_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of pointer::offset.
+/// - Call `zcashlc_free_boxed_slice` to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_derive_arbitrary_wallet_key(
+    context_string: *const u8,
+    context_string_len: usize,
+    seed: *const u8,
+    seed_len: usize,
+) -> *mut FfiBoxedSlice {
+    let res = catch_panic(|| {
+        let context_string = unsafe { slice::from_raw_parts(context_string, context_string_len) };
+        let seed = unsafe { slice::from_raw_parts(seed, seed_len) };
+
+        let key = arbitrary::SecretKey::from_path(context_string, seed, &[]);
+
+        Ok(FfiBoxedSlice::some(key.data().to_vec()))
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the account level.
+///
+/// `context_string` is a globally-unique non-empty sequence of at most 252 bytes that identifies
+/// the desired context.
```

Comment:
Similar to `zcashlc_derive_arbitrary_wallet_key`, either document the panic or make it an error.